### PR TITLE
FeignException uses request and response but doesn't use a response body

### PR DESCRIPTION
**Pull request by radio-rogal**
Simplifying of FeignException: it uses request and response objects but does not use a response body as byte array.

It is breaking changes if any extension use the methods _content_. _contentUTF8_, _responseBody_ or _responseHeaders_.
FeignException and DecodeException have new constructors.

These changers start from the discussion #1452

P.S.
I like [the idea to make FeignException more simplier](https://github.com/OpenFeign/feign/pull/1452#issuecomment-875538393). But I have not found how to do it without breaking changes.

### CODE REVIEW

File: core/src/main/java/feign/FeignException.java [hunk 1]
```diff
@@ -37,108 +37,61 @@
...
+  private static final String EXCEPTION_MESSAGE_TEMPLATE_NULL_RESPONSE =
+      "response should not be null";
   private static final long serialVersionUID = 0;
-  private int status;
-  private byte[] responseBody;
-  private Map<String, Collection<String>> responseHeaders;
+  private final int status;
   private Request request;
-
-  protected FeignException(int status, String message, Throwable cause) {
```
Comment (velo) [inline core/src/main/java/feign/FeignException.java:46]: I HATE this class!

That said, it's not safe to remove the extra constructors, as people may be using it.

Bring them back, and feel free to deprecate them all

File: core/src/main/java/feign/FeignException.java [hunk 2]
```diff
@@ -37,108 +37,61 @@
...
-  public byte[] content() {
-    return this.responseBody;
-  }
-
-  /**
-   * The Response body.
-   *
-   * @return an Optional wrapping the response body.
-   */
-  public Optional<ByteBuffer> responseBody() {
```
Comment (velo) [inline core/src/main/java/feign/FeignException.java:130]: as with the original constructors... we need to keep the old methods for compatibility.

Please bring them all back, not sure if they got deleted or just sorted, keeping them on their original position makes this more reviewable.

File: core/src/main/java/feign/FeignException.java [hunk 3]
```diff
@@ -271,149 +203,130 @@ static FeignException errorExecuting(Request request, IOException cause) {
   }
 
   public static class FeignClientException extends FeignException {
-    public FeignClientException(int status, String message, Request request, byte[] body,
```
Comment (velo) [inline core/src/main/java/feign/FeignException.java:274]: anything public, needs to be kept, add `@deprected` to them, but keep them in place

**Review by velo — CHANGES_REQUESTED** [review]
Overall I liked the change, but, due to feign nature, we CAN'T just remove public/protected methods/constructors.  At least not in minor releases.

Please bring them back, mark as deprecated and we may nuke on feign 12...

**Comment by velo** [comment]
> I like [the idea to make FeignException more simplier](https://github.com/OpenFeign/feign/pull/1452#issuecomment-875538393). But I have not found how to do it without breaking changes.

I feel ya

File: core/src/main/java/feign/FeignException.java (hunk 2 above)
Comment (radio-rogal) [reply core/src/main/java/feign/FeignException.java:130]: done in the commit 2523ef1

File: core/src/main/java/feign/FeignException.java (hunk 1 above)
Comment (radio-rogal) [reply core/src/main/java/feign/FeignException.java:46]: done in the commit 2523ef1

File: core/src/main/java/feign/FeignException.java (hunk 3 above)
Comment (radio-rogal) [reply core/src/main/java/feign/FeignException.java:274]: done in the commit 2523ef1

**Comment by radio-rogal** [comment]
The Snyk upgraded jackson's version in c88b65a and the build was successful. But then I update my branch and it fell for jdk14 only.
Unfortunatelly I can see what is wrong.

**Comment by radio-rogal** [comment]
Today I tried to return unit-tests for response body methods. I thought that all these methods should be deprecated. But I found that they are required as a response could have a non-repeatable body so:

-  we could not made an exception message without touching a response body
-  we could not get a response body from exception if a response has non-repeatable body as we just read it when created new exception.

**Comment by radio-rogal** [comment]
It requires a different approach, perhaps you need to transfer everything to factory methods.

**Pull request closed without merging**